<hr>

# Assigment: Recurrence Relation

**Author:** *Ritesh Mandal*

**Submission Date:** 30-11-23

**Module:** Data Structure and Algorithms

**Instructor:** Priya Bhatia

**Python version:** 3.11.4

<hr>

<hr>

### Q1. Find the value of T(2), given that `T(0)=5` for the recurrence relation 
### &emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp; `T(n) = 3T(n-1) + 12n`.
### Sol: 

In [42]:
def recurrence_relation_1(n):
    if n == 0:
        return 5
    return 3 * recurrence_relation_1(n-1) + 12 * n

n = 2
print(f"T({n}) = {recurrence_relation_1(n)}")

T(2) = 105


1. Substituting `n = 1` in `T(n) = 3T(n-1) + 12n`

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \boldsymbol{T(\textcolor{green}{1})} = 3T(\textcolor{green}{1} -1) + 12 \times \textcolor{green}{1}}$

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \boldsymbol{T(1)} = 3T(0) + 12}$

&emsp;&emsp;Given **`T(0) = 5`**

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \boldsymbol{T(1)} = 3 \times \textcolor{green}{5} + 12}$

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \boldsymbol{T(1)} = \textcolor{green}{15} + 12}$

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\implies \boxed{\boldsymbol{T(1)} = \textcolor{green}{27}}}$

2. Now substituting `n = 2` in `T(n) = 3T(n-1) + 12n`

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \boldsymbol{T(\textcolor{green}{2})} = 3T(\textcolor{green}{2} -1) + 12 \times \textcolor{green}{2}}$

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \boldsymbol{T(2)} = 3T\color{green}{(1)}+ 24}$

&emsp;&emsp;Substituting, **`T(1) = 27`**

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \boldsymbol{T(2)} = 3 \times \textcolor{green}{27}  + 24}$

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow\boldsymbol{T(2)} = \textcolor{green}{81}  + 24}$

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\implies \color{green}{\boxed{\boldsymbol{T(2)} = \textcolor{green}{105}}}}$

<hr>

<hr>

### Q2. Given a recurrence relation, solve it using the substitution method:

#### a.&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;`T(n) = T(n-1) + c`

### Sol:

In [35]:
def recurrence_relation_2a(n,c):
    if n == 1:
        return 1 
    return recurrence_relation_2a(n-1, c) + c

n = int(input("Enter the value of n: "))
C = 10
print(f"T({n}) = {recurrence_relation_2a(n,C)}")

T(10) = 91


&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\boxed{\boldsymbol{T(n)} = T(n-1) + c}} \;\rightarrow eqn1$

1. Substituting `n = n - 1` in `eqn1`

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \boldsymbol{T(\textcolor{green}{n - 1})} = T(\textcolor{green}{(n - 1)} -1) + c}$

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \color{green}{\boxed{\boldsymbol{T(n - 1)} = T(n -2) + c}}} \;\rightarrow eqn2$

2. Substituting the value of `T(n-1)` from `eqn2` in `eqn1`

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \boldsymbol{T(n)} = [\; T(\textcolor{green}{(n - 2)}) + c\;] + c}$

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \color{green}{\boxed{\boldsymbol{T(n)} = T(n -2) + 2c}}} \;\rightarrow eqn3$

3. Substituting `n = n - 2` in `eqn1`

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \boldsymbol{T(\textcolor{green}{n - 2})} = T(\textcolor{green}{(n - 2)} -1) + c}$

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \color{green}{\boxed{\boldsymbol{T(n - 2)} = T(n -3) + c}}} \;\rightarrow eqn4$

4. Substituting the value of `T(n-2)` from `eqn4` in `eqn3`

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \boldsymbol{T(n)} = [\; T(\textcolor{green}{(n - 3)}) + c\;] + 2c}$

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \color{green}{\boxed{\boldsymbol{T(n)} = T(n -3) + 3c}}} \;\rightarrow eqn5$

5. Generalising for `k` times we get,

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \color{green}{\boxed{\boldsymbol{T(n)} = T(n -k) + k.c}}} \;\rightarrow eqn6$

6. Substituting `k = n-1` in `eqn6` i.e., repeating the operation `n-1` times, we get 

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow {\boldsymbol{T(n)} = T(n - \color{green}{(n-1)}\color{black}{)} + \color{green}{(n-1)}.c}}$

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\implies  \color{green}{\boxed{\boldsymbol{T(n)} = T(1) + (n-1).c}}}$ $\mathrm{\equiv \color{green}{\boxed{\boldsymbol{\bigcirc (n)}}}}$

<hr>

### b.&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;`T(n) = 2T(n/2) + n`
### Sol:

In [41]:
def recurrence_relation_2b(n):
    if n == 1:
        return 1
    return 2 * recurrence_relation_2b(n//2) + n

n = int(input("Enter the value of n: "))
print(f"T({n}) = {recurrence_relation_2b(n)}")

T(8) = 32


&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\boxed{\boldsymbol{T(n)} = 2T\left(\frac{n}{2}\right) + n}} \;\rightarrow eqn1$

1. Substituting `n = n/2` in `eqn1`

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \boxed{\boldsymbol{T\textcolor{green}{\left(\frac{n}{2}\right)}} = 2T\color{green}{\left(\frac{n}{2^2}\right)}+ \color{green}{\left(\frac{n}{2}\right)}}}\;\rightarrow eqn2$

2. Substituting the value of `T(n/2)` from `eqn2` in `eqn1`

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \boldsymbol{T(n)} = 2 \times[\; \color{green}{2T\left(\frac{n}{2^2}\right)+ \left(\frac{n}{2}\right)}}\;] + n$

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \boldsymbol{T(n)} = 2^2.T\left(\frac{n}{2^2}\right)+ 2.\left(\frac{n}{2}\right) + n}$

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \boxed{\boldsymbol{T(n)} = 2^2.T\left(\frac{n}{2^2}\right)+ \color{green}{2n}}}\;\rightarrow eqn3$

3. Substituting `n = n/4` in `eqn1`

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \boxed{\boldsymbol{T\textcolor{green}{\left(\frac{n}{2^2}\right)}} = 2T\color{green}{\left(\frac{n}{2^3}\right)}+ \color{green}{\left(\frac{n}{2^2}\right)}}}\;\rightarrow eqn4$

4. Substituting the value of `T(n/4)` from `eqn4` in `eqn3`

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \boldsymbol{T(n)} = 2^2\times[\; \color{green}{2T\left(\frac{n}{2^3}\right)+ \left(\frac{n}{2^2}\right)}}\;] + 2n$

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \boldsymbol{T(n)} = 2^3.T\left(\frac{n}{2^3}\right)+ 2^2.\left(\frac{n}{2^2}\right) + 2n}$

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \boxed{\boldsymbol{T(n)} = 2^3.T\left(\frac{n}{2^3}\right)+ \color{green}{3n}}}\;\rightarrow eqn5$

5. Generalising for `k` times we get,

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \color{green}{\boxed{\boldsymbol{T(n)} = 2^k.T\left(\frac{n}{2^k}\right)+ k.n}}}\;\rightarrow eqn6$

6. Substituting $\mathrm{\left(\frac{n}{2^k}\right) = 1, \; \text{i.e.,}\; 2^k = n \implies \boxed{k = \log_2n}}$ in `eqn6`

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \boldsymbol{T(n)} = 2^{\color{green}{\log_2{n}}}.T\left(\frac{n}{2^{\color{green}{\log_2{n}}}}\right)+ ({\color{green}{\log_2{n}}})n}$

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \boldsymbol{T(n)} = {\color{green}{n}}.T\left(\frac{n}{{\color{green}{n}}}\right)+ n.{\log_2{n}}}$

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\implies {\color{green}{\boxed{\boldsymbol{T(n)} = n.T(1)+ n.{\log_2{n}}}}}}$ $\mathrm{\equiv \color{green}{\boxed{\boldsymbol{\bigcirc (n.logn)}}}}$

<hr>

### c.&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;`T(n) = 2T(n/2) + c`
### Sol:

In [36]:
def recurrence_relation_2c(n,c):
    if n == 1:
        return 1
    return 2 * recurrence_relation_2c(n//2, c) + c

n = int(input("Enter the value of n: "))
C = 10
print(f"T({n}) = {recurrence_relation_2c(n,C)}")

T(4) = 34


&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\boxed{\boldsymbol{T(n)} = 2T\left(\frac{n}{2}\right) + c}} \;\rightarrow eqn1$

1. Substituting `n=n/2`​ in `eqn1`

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \boxed{\boldsymbol{T\textcolor{green}{\left(\frac{n}{2}\right)}} = 2T{\color{green}{\left(\frac{n}{2^2}\right)}}+ c}}\;\rightarrow eqn2$

2. Substituting the value of `T(n/2)` from `eqn2` in `eqn1`

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \boldsymbol{T(n)} = 2 \times[\; \color{green}{2T\left(\frac{n}{2^2}\right)+ c}}\;] + c$

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \boldsymbol{T(n)} = 2^2.T\left(\frac{n}{2^2}\right)+ 2c + c}$

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \boxed{\boldsymbol{T(n)} = 2^2.T\left(\frac{n}{2^2}\right)+ \color{green}{3c}}}\;\rightarrow eqn3$

3. Substituting `n = n/4` in `eqn1`

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \boxed{\boldsymbol{T\textcolor{green}{\left(\frac{n}{2^2}\right)}} = 2T{\color{green}{\left(\frac{n}{2^3}\right)}}+ c}}\;\rightarrow eqn4$

4. Substituting the value of `T(n/4)` from `eqn4` in `eqn3`

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \boldsymbol{T(n)} = 2^2\times[\; {\color{green}{2T\left(\frac{n}{2^3}\right)+ c}}\;] + 3c}$

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \boldsymbol{T(n)} = 2^3.T\left(\frac{n}{2^3}\right)+ {\color{green}{4c}} + 3c}$

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \boxed{\boldsymbol{T(n)} = 2^3.T\left(\frac{n}{2^3}\right)+ \color{green}{7c}}}\;\rightarrow eqn5$

5. Generalising for `k` times we get,

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \color{green}{\boxed{\boldsymbol{T(n)} = 2^k.T\left(\frac{n}{2^k}\right)+ (2^k-1)c}}}\;\rightarrow eqn6$

6. Substituting $\mathrm{\left(\frac{n}{2^k}\right) = 1, \; \text{i.e.,}\; 2^k = n \implies \boxed{k = \log_2n}}$ in `eqn6`

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \boldsymbol{T(n)} = 2^{\color{green}{\log_2{n}}}.T\left(\frac{n}{2^{\color{green}{\log_2{n}}}}\right)+ (2^{\color{green}{{\log_2{n}}}}-1)c}$

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \boldsymbol{T(n)} = {\color{green}{n}}.T\left(\frac{n}{{\color{green}{n}}}\right)+ ({\color{green}{n}}-1)c}$

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\implies {\color{green}{\boxed{\boldsymbol{T(n)} = n.T(1)+ (n-1)c}}}}$ $\mathrm{\equiv \color{green}{\boxed{\boldsymbol{\bigcirc (n)}}}}$

<hr>

### d.&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;`T(n) = T(n/2) + c`
### Sol:

In [38]:
def recurrence_relation_2d(n,c):
    if n == 1:
        return 1
    return recurrence_relation_2d(n//2, c) + c

n = int(input("Enter the value of n: "))
C = 10
print(f"T({n}) = {recurrence_relation_2d(n,C)}")

T(8) = 31


&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\boxed{\boldsymbol{T(n)} = T\left(\frac{n}{2}\right) + c}} \;\rightarrow eqn1$

1. Substituting `n=n/2`​ in `eqn1`

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \boxed{\boldsymbol{T\textcolor{green}{\left(\frac{n}{2}\right)}} = T{\color{green}{\left(\frac{n}{2^2}\right)}}+ c}}\;\rightarrow eqn2$

2. Substituting the value of `T(n/2)` from `eqn2` in `eqn1`

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \boldsymbol{T(n)} = [\; \color{green}{T\left(\frac{n}{2^2}\right)+ c}}\;] + c$

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \boldsymbol{T(n)} = T\left(\frac{n}{2^2}\right)+ c + c}$

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \boxed{\boldsymbol{T(n)} = T\left(\frac{n}{2^2}\right)+ \color{green}{2c}}}\;\rightarrow eqn3$

3. Substituting `n = n/4` in `eqn1`

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \boxed{\boldsymbol{T\textcolor{green}{\left(\frac{n}{2^2}\right)}} = T{\color{green}{\left(\frac{n}{2^3}\right)}}+ c}}\;\rightarrow eqn4$

4. Substituting the value of `T(n/4)` from `eqn4` in `eqn3`

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \boldsymbol{T(n)} = [\; {\color{green}{T\left(\frac{n}{2^3}\right)+ c}}\;] + 2c}$

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \boxed{\boldsymbol{T(n)} = T\left(\frac{n}{2^3}\right)+ \color{green}{3c}}}\;\rightarrow eqn5$

5. Generalising for `k` times we get,

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \color{green}{\boxed{\boldsymbol{T(n)} = T\left(\frac{n}{2^k}\right)+ k.c}}}\;\rightarrow eqn6$

6. Substituting $\mathrm{\left(\frac{n}{2^k}\right) = 1, \; \text{i.e.,}\; 2^k = n \implies \boxed{k = \log_2n}}$ in `eqn6`

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \boldsymbol{T(n)} = T\left(\frac{n}{2^{\color{green}{\log_2{n}}}}\right)+ ({\color{green}{{\log_2{n}}}})c}$

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\Rightarrow \boldsymbol{T(n)} = T\left(\frac{n}{{\color{green}{n}}}\right)+ c.\log_2{n}}$

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\implies {\color{green}{\boxed{\boldsymbol{T(n)} = T(1)+ c.\log_2{n}}}}}$ $\mathrm{\equiv \color{green}{\boxed{\boldsymbol{\bigcirc (logn)}}}}$

<hr>

<hr>

### Q3. Given a recurrence relation, solve it using the recursive tree approach:

### a.&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;`T(n) = 2T(n-1) +1`
### Sol:
                        T(n)    --> at the root level (level 0), we have extra 1 cost.
                       /    \
                 T(n-1)     T(n-1)    --> at level 1, we have 1 + 1 = 2 extra cost.
                 /   \       /    \
           T(n-2) T(n-2) T(n-2)  T(n-2)    --> at level 2, we have 4 * 1 =  4 extra cost.
            / \     / \    / \     / \
       T(n-3) T(n-3) ... (repeated) T(n-3)    --> at level 3, we have  8 * 1 = 8 extra cost.
         / \    / \                /    \
    T(n-4) T(n-4) ... (repeated)  T(n-4)  ... and so on
      / \      / \
      
      --- k-th level ---
    T(n-k) T(n-k) ... (repeated)  T(n-k)  ...   T(n-k)  

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\implies\text{at level k, we have } 2^k \times 1 = \color{green}{2^k\text{ }extra\text{ }cost}}$

$\mathrm{\text{If n - k = 1 }\implies \text{at }k^{th}\text{ level, we reach the base case condition of }\color{green}{\boldsymbol{T(1)}}}$

$\mathrm{\implies\color{green}{\boldsymbol{k = n-1}}}$ i.e., the depth of recursive tree is **level 0** to **level (n-1)**

$\mathrm{\text{Hence, }T(n) = 1 + 2 + 2^2 + 2^3 + ........... + 2^{n-1}}$

$\mathrm{\implies T(n) = \left(\frac{1.(2^n - 1)}{2\text{ }-\text{ }1}\right)}$

$\mathrm{\implies \color{green}{\boxed{\boldsymbol{T(n) = 2^n - 1}}}}$ $\mathrm{\equiv \color{green}{\boxed{\boldsymbol{\bigcirc (2^n)}}}}$

In [39]:
def recurrence_relation_3a(n):
    if n == 1:
        return 1
    return 2 * recurrence_relation_3a(n-1) + 1

n = int(input("Enter the value of n: "))
print(f"T({n}) = {recurrence_relation_3a(n)}")

T(4) = 15


<hr>

### b.&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;`T(n) = 2T(n/2) + n`
### Sol:
                        T(n)    --> at the root level (level 0), we have extra n cost.
                       /    \
                 T(n/2)     T(n/2)    --> at level 1, we have n/2 + n/2 = n extra cost.
                 /   \       /    \
           T(n/4) T(n/4) T(n/4)  T(n/4)    --> at level 2, we have 4 * n/4 =  n extra cost.
            / \     / \    / \     / \
       T(n/8) T(n/8) ... (repeated) T(n/8)    --> at level 3, we have  8 * n/8 = n extra cost.
         / \    / \                /    \
    T(n/16) T(n/16) ... (repeated)  T(n/16)  ... and so on
      / \      / \
      
      --- k-th level ---
    T(n/2^k) T(n/2^k) ... (repeated)  T(n/2^k)  

&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;$\mathrm{\implies\text{at level k, we have } 2^k \times \left(\frac{n}{2^k}\right) = \color{green}{\boldsymbol{n}\text{ }extra\text{ }cost}}$

$\mathrm{\text{If }\left(\frac{n}{2^k}\right) = 1\implies \text{at }k^{th}\text{ level, we reach the base case condition of }\color{green}{\boldsymbol{T(1)}}}$

$\mathrm{\implies 2^k = n}$

$\mathrm{\implies\color{green}{\boldsymbol{k = log_2{n}}}}$ i.e., the depth of recursive tree is from **level 0** to **level** ( $\boldsymbol{log_2{n}}$ )

$\mathrm{\text{Hence, }T(n) = n + n + n + n + .....+n \text{    } \rightarrow(log_2{n} + 1) times}$

$\mathrm{\implies T(n) = (log_2{n} + 1)\times n}$

$\mathrm{\implies \color{green}{\boxed{\boldsymbol{T(n) = n.(log_2{n}+ 1)}}}}$ $\mathrm{\equiv \color{green}{\boxed{\boldsymbol{\bigcirc (n.logn)}}}}$

In [40]:
def recurrence_relation_3b(n):
    if n == 1:
        return 1
    return 2 * recurrence_relation_3b(n//2) + n

n = int(input("Enter the value of n: "))
print(f"T({n}) = {recurrence_relation_3b(n)}")

T(4) = 12
